src/components/vm/nics/nicBody.tsx

```diff
@@ -271,3 +290,311 @@ export const NetworkTypeAndSourceRow = ({
         </>
     );
 };
+
+export function portForwardText(pf: VMInterfacePortForward): string {
+    let text = "";
+
+    if (pf.address)
+        text += pf.address;
+    if (pf.dev)
+        text += pf.dev;
+    if (!pf.range.some(r => r.exclude != "yes")) {
+        if (text != "")
+            text += ", ";
+        text += _("all ports");
+    }
+    for (let i = 0; i < pf.range.length; i++) {
+        const r = pf.range[i];
+        if (text != "")
+            text += ", ";
+        if (r.exclude == "yes")
+            text += _("excluding ");
+        text += r.start;
+        if (r.end)
+            text += "-" + r.end;
+        if (r.to)
+            text += " → " + r.to + (r.end ? "-" + (Number(r.to) + Number(r.end) - Number(r.start)).toString() : "");
+        if (pf.proto && pf.proto != "tcp")
+            text += "/" + pf.proto;
+    }
+
+    return text;
+}
+
+export function interfacePortForwardsToDialog(portForwards: VMInterfacePortForward[]): DialogPortForward[] {
+    return portForwards.map(pf => {
+        if (!pf.dev && pf.range.length == 1 && pf.range[0].exclude != "yes") {
+            const r = pf.range[0];
+            return {
+                kind: "simple",
+                address: pf.address || "",
+                proto: pf.proto || "",
+                host: r.start + (r.end ? "-" + r.end : ""),
+                guest: r.to || "",
+            };
+        } else {
+            return {
+                kind: "complex",
+                config: pf,
+            };
+        }
+    });
+}
+
+export function dialogPortForwardsToInterface(portForwards: DialogPortForward[]): VMInterfacePortForward[] {
+    // NOTE - Instead of validating the input (which no other part of
+    // the dialog does), we simply ignore rules with a empty host
+    // port.
+
+    function valid_pf(pf: DialogPortForward): boolean {
+        return !(pf.kind == "simple" && !pf.host);
+    }
+
+    return portForwards.filter(valid_pf).map((pf: DialogPortForward) => {
+        if (pf.kind == "simple") {
+            const range = pf.host.split("-");
+            return {
+                address: pf.address || null,
+                dev: null,
+                proto: pf.proto,
+                range: [
+                    {
+                        start: range[0],
+                        end: range[1] || null,
+                        to: pf.guest || null,
+                        exclude: null,
+                    }
+                ],
+            };
+        } else {
+            return pf.config;
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2331-3b8315e1-20251022-060022-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)